src/Orleans.Runtime/Silo/Silo.cs

```diff
@@ -654,21 +671,22 @@ public void Stop()
         /// </summary>
         public void Shutdown()
         {
-            Terminate(true);
+            this.gracefully = true;
```

Comment:
I'm not certain about this, but does it make sense to use a cancelled CTS instead of having a `gracefully` field to communicate ungraceful termination? On one hand, it means that ungraceful termination is communicated to all lifecycle subscribers, but on the other they might not know how to deal with that and might throw `OperationCancelledException`. It's just a thought, you probably already considered it. In the ASP.NET Generic Host case, they use a cancelled token to signify ungraceful termination.

Reply:
I'm not sure, unfortunately. :/  We should think about this more.

In the lifecycle, a canceled token means cancel the operation.  So on Start it means cancel the start operation and on Stop it means cancel the stop operation.  This means that, when canceled, the lifecycle observers never even get called, so they never get a chance to even perform an ungraceful shutdown.  I think this is the desired behavior, but I've not fully thought through the implications of this pattern in the context of existing shutdown logic.  The goal of this PR was to wrap the existing behaviors in the lifecycle pattern (which I think it does..?) allowing us to refactor over time to improve the system.

In general, I think a canceled token should indicate ungraceful shutdown, but this should mean that lifecycle observers are simply not called to shutdown, not that they are called in a way that let's them know they are being ungracefully shutdown.  IMO, we should consider how to tell components to ungracefully shutdown only if the simpler pattern demonstrates itself as insufficient.  This means removing the 'gracefully' flag and related logic in a future refactor.

Reply:
I'm fine with this, let's keep the bool for now

Reply:
I don't think the current approach is good, as it only works because we temporarily made the `Silo` itself participate in the lifecycle, and it is the only one that knows whether it's graceful or ungraceful shutdown.
I think we should allow every single participant know whether the shutdown is graceful or not and let *it* decide to no-op if it's not important. Unfortunately that means querying the cancelation token, which is a little unintuitive, but can be documented and eventually will be familiar to people creating `IHostedService` implementations too (and that's a good thing, as Orleans extensibility will be similar to the extensibility of the generic Host).
Canceling Start IMO should preclude more participating services to start, and the currently starting services should receive the cancellation request. Then (externally the Silo will) initiate the reverse stopping from that stage on, with a canceled token or not, depending whether the shutdown is graceful.
Also, the host might initiate a graceful stop, but a few seconds in might decide to cancel the token and continue as ungraceful (this is to note that the "Start" and "Stop" part of the lifecycle use effectively different cancellation tokens).

For more info about the semantics of IHostedService, please see the comments in there: [IHostedService](https://github.com/aspnet/Hosting/blob/dev/src/Microsoft.Extensions.Hosting.Abstractions/IHostedService.cs)
BTW, looking at [how they are executed](https://github.com/aspnet/Hosting/blob/dev/src/Microsoft.Extensions.Hosting/Internal/Host.cs#L51), it looks like even if the Start operation is canceled, they still call Start on all the IHostServices implementations, but passing a canceled token. I'm fine with doing that or not even starting the remainder of participants, since I can't think of anything that a lifecycle participant could ever do (other than no-op) on Start if the token is canceled. But I'm fine either way.

Reply:
What you suggest makes sense and is more inline with asp.net.  On further reflection, 'canceling' the stop operation does not make as much sense as canceling the start operation, so using the token to communicate graceful/ungraceful shutdown has more utility.  If we find we need to actually cancel a stop operation, we can address that when needed.

Reply:
On hindsight, even the Shutdown operation can pass a token that automatically cancels after 5 minutes, in case there is something stalling the shutdown

Reply:
Let's note these concerns and resolve them in design discussions, not in PR.  If this PR preserves the existing behaviors and enables future improvements, then it has achieved it's purpose.